iree/compiler/Dialect/Flow/Transforms/DispatchLinalgOnTensors.cpp

```diff
@@ -0,0 +1,309 @@
+    }
+    // TODO(nicolasvasilache): return `clonedRoot->getResults()` once we have
+    // shape operands and we drop tie_shape.
+    b.create<IREE::Flow::ReturnOp>(loc);
+  }
+  return dispatchOp;
+}
+
+// Rewrite pattern to ensure only ops with tensor semantics are tiled.
+struct TileAndDistributeOnTensorsPattern
+    : public linalg::LinalgBaseTilingPattern {
+  using Base = linalg::LinalgBaseTilingPattern;
+  TileAndDistributeOnTensorsPattern(linalg::LinalgTilingOptions options,
+                                    linalg::LinalgMarker marker,
+                                    PatternBenefit benefit = 1)
+      : Base(options, marker, benefit) {}
+
+  LogicalResult matchAndRewrite(Operation *op,
+                                PatternRewriter &rewriter) const override {
+    auto linalgOp = dyn_cast<linalg::LinalgOp>(op);
+    if (!linalgOp || !linalgOp.hasTensorSemantics()) return failure();
+
+    linalg::LinalgOp clonedLinalgOp;
+    Operation *dispatch =
+        buildFlowWorkgroupDispatchOp(rewriter, linalgOp, clonedLinalgOp);
+
+    OpBuilder::InsertionGuard g(rewriter);
+    rewriter.setInsertionPoint(clonedLinalgOp);
+    SmallVector<Value, 4> tensorResults;
+    if (failed(Base::matchAndRewriteBase(clonedLinalgOp, rewriter,
+                                         tensorResults))) {
+      // Need to erase on failure to avoid infinite loop
+      rewriter.eraseOp(dispatch);
+      return failure();
+    }
+    rewriter.replaceOp(op, dispatch->getResults());
+    rewriter.replaceOp(clonedLinalgOp, tensorResults);
+
+    return success();
+  }
+};
+
+static Value buildFlowWorkgroupIdOp(OpBuilder &b, unsigned dim) {
+  return b.create<IREE::Flow::DispatchWorkgroupIDOp>(
+      b.getInsertionPoint()->getLoc(), dim);
+}
+
+static Value buildFlowWorkgroupCountOp(OpBuilder &b, unsigned dim) {
+  return b.create<IREE::Flow::DispatchWorkgroupCountOp>(
+      b.getInsertionPoint()->getLoc(), dim);
+}
+
+// After outlining in dispatch region we can rewrite the dispatch ops with
+// proper captures.
+static void setDispatchWorkgroupOperands(
+    IREE::Flow::DispatchWorkgroupsOp dispatchOp) {
+  llvm::SetVector<Value> valuesSet;
+  mlir::getUsedValuesDefinedAbove(dispatchOp.body(), valuesSet);
+  ValueRange valuesDefinedAbove{valuesSet.getArrayRef()};
+
+  SmallVector<Value, 4> originalDispatchOperands = dispatchOp.operands();
+  llvm::SetVector<Value> originalDispatchOperandsSet(
+      originalDispatchOperands.begin(), originalDispatchOperands.end());
+  SmallVector<Value, 4> newOperands, clones;
+  for (Value v : valuesDefinedAbove) {
+    if (v.getDefiningOp<ConstantOp>()) {
+      clones.push_back(v);
+    } else if (!originalDispatchOperandsSet.contains(v)) {
+      newOperands.push_back(v);
+    }
+  }
+  // Need to insert new operands into operands and BB args at proper
+  // positions.
+  assert(newOperands.empty() && "unsupported new operands from capture");
```

Comment:
Not really sure how we can be sure everything needed has been pulled into dispatch region. This probably works only for the single op case handled here, but I am not sure this will work when we have multiple ops.

Reply:
That's why the assertion is here, this pass does not claim to support all use cases.
The whole PR is more general than making this pass work for all cases: learn to walk before we can run and all.